File: unity/Assets/Scripts/controller/MockMotionController.cs
// MockMotionController — viewer/src/controller/MockMotionController.ts의 C# 1:1 이식.
// 단위 테스트 / parity test용 spy. timer 없음 — 모든 lifecycle은 명시적 호출로만.
// 테스트가 Settle/Stop을 직접 부르고 emit된 event 시퀀스를 검증.

#nullable enable

using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Nubjuk.Viewer.Types;
using Nubjuk.Viewer.Ws;

namespace Nubjuk.Viewer.Controller
{
    public sealed class MockMotionController : IMotionController
    {
        private MockActive? _active;
        private readonly List<Action<MotionEvent>> _handlers = new();
        private readonly List<MotionEvent> _events = new();
        private readonly List<MotionSettleResult> _settleCalls = new();

        public IReadOnlyList<MotionEvent> GetEvents() => _events;
        public IReadOnlyList<MotionSettleResult> GetSettleCalls() => _settleCalls;

        public void ResetSpy()
        {
            _active = null;
            _events.Clear();
            _settleCalls.Clear();
        }

        public Task PlayAsync(IntentName intent, long durationMs, string correlationId)
        {
            if (_active != null && _active.CorrelationId == correlationId)
            {
                return Task.CompletedTask;
            }
            if (_active != null)
            {
                var prev = _active;
                _active = null;
                Emit(new MotionEvent.Cancelled(prev.CorrelationId, CancelReason.Superseded));
            }
            _active = new MockActive(intent, durationMs, correlationId, DateTimeOffset.UtcNow.ToUnixTimeMilliseconds());
            Emit(new MotionEvent.Started(correlationId, intent, durationMs));
            return Task.CompletedTask;
        }

        public void Settle(MotionSettleResult result)
        {
            // 호출 자체 기록 (mismatched cid no-op도 포함) — dispatcher 회귀 가드.
            _settleCalls.Add(result);
            if (_active == null || _active.CorrelationId != result.CorrelationId) return;
            var cid = _active.CorrelationId;
            _active = null;
            switch (result)
            {
                case MotionSettleResult.Completed completed:
                    Emit(new MotionEvent.Completed(cid, completed.ActualMs));
                    break;
                case MotionSettleResult.Failed failed:
                    Emit(new MotionEvent.Failed(cid, failed.Reason));
                    break;
            }
        }

        public void Stop()
        {
            if (_active == null) return;
            var prev = _active;
            _active = null;
            Emit(new MotionEvent.Cancelled(prev.CorrelationId, CancelReason.StopCalled));
        }

        public bool IsPlaying() => _active != null;
        public string? GetCurrentCorrelationId() => _active?.CorrelationId;

        public ISubscription OnEvent(Action<MotionEvent> handler)
        {
            _handlers.Add(handler);
            return new SubscriptionImpl(() => _handlers.Remove(handler));
        }

        private void Emit(MotionEvent ev)
        {
            _events.Add(ev);
            // 핸들러 변경 중 호출 방지를 위해 snapshot 순회
            var snapshot = _handlers.ToArray();
            foreach (var h in snapshot) h(ev);
        }

        private sealed record MockActive(IntentName Intent, long DurationMs, string CorrelationId, long StartedAtMs);

        private sealed class SubscriptionImpl : ISubscription
        {
            private readonly Action _dispose;
            private bool _disposed;
            public SubscriptionImpl(Action dispose) { _dispose = dispose; }
            public void Dispose()
            {
                if (_disposed) return;
                _disposed = true;
                _dispose();
            }
        }
    }
}

File: src/controller/MockMotionController.ts
// MockMotionController — 단위 테스트용 spy.
// MotionController 인터페이스 만족. 자체 timer 없음 — 모든 lifecycle은 명시적 호출로만.
// 테스트가 settle/stop을 직접 호출하고 emit된 event 시퀀스를 검증.

import type { IntentName } from "@/types/protocol";
import type { Subscription } from "@/ws/ViewerConnection";
import type { MotionController, MotionEvent, MotionSettleResult } from "./MotionController";

interface MockActive {
  intent: IntentName;
  durationMs: number;
  correlationId: string;
  startedAt: number;
}

export class MockMotionController implements MotionController {
  private active: MockActive | null = null;
  private handlers = new Set<(ev: MotionEvent) => void>();
  private events: MotionEvent[] = [];
  // dispatcher가 settle을 부른 *호출 자체* 추적 — cid 불일치로 no-op 된 호출도 포함.
  // 테스트가 "stale 메시지에 settle 호출 자체가 안 됐는지" 검증할 때 사용 (Codex P2).
  private settleCalls: MotionSettleResult[] = [];

  // 테스트가 검증하는 emit log
  getEvents(): readonly MotionEvent[] {
    return this.events;
  }

  // 테스트가 settle 호출 시도 자체를 검증 — events와 달리 mismatched cid도 기록.
  getSettleCalls(): readonly MotionSettleResult[] {
    return this.settleCalls;
  }

  resetSpy(): void {
    this.active = null;
    this.events = [];
    this.settleCalls = [];
  }

  async play(intent: IntentName, durationMs: number, correlationId: string): Promise<void> {
    if (this.active?.correlationId === correlationId) return;
    if (this.active) {
      const prev = this.active;
      this.active = null;
      this.emit({
        type: "cancelled",
        correlationId: prev.correlationId,
        reason: "superseded",
      });
    }
    this.active = {
      intent,
      durationMs,
      correlationId,
      startedAt: performance.now(),
    };
    this.emit({ type: "started", correlationId, intent, expectedMs: durationMs });
  }

  settle(result: MotionSettleResult): void {
    // 호출 자체를 기록 (cid 불일치로 no-op 되더라도) — dispatcher 회귀 가드용.
    this.settleCalls.push(result);
    if (!this.active || this.active.correlationId !== result.correlationId) return;
    const cid = this.active.correlationId;
    this.active = null;
    if (result.type === "completed") {
      this.emit({ type: "completed", correlationId: cid, actualMs: result.actualMs });
    } else {
      this.emit({ type: "failed", correlationId: cid, reason: result.reason });
    }
  }

  stop(): void {
    if (!this.active) return;
    const prev = this.active;
    this.active = null;
    this.emit({
      type: "cancelled",
      correlationId: prev.correlationId,
      reason: "stop_called",
    });
  }

  isPlaying(): boolean {
    return this.active !== null;
  }

  getCurrentCorrelationId(): string | null {
    return this.active?.correlationId ?? null;
  }

  onEvent(handler: (ev: MotionEvent) => void): Subscription {
    this.handlers.add(handler);
    return {
      dispose: () => {
        this.handlers.delete(handler);
      },
    };
  }

  private emit(ev: MotionEvent) {
    this.events.push(ev);
    this.handlers.forEach((h) => h(ev));
  }
}
